lib/Service/SMimeService.php

```diff
@@ -65,4 +71,64 @@ public function verifyMessage(string $message): bool {
 
 		return $valid;
 	}
+
+	/**
+	 * Parse a X509 certificate and extract the email address from its subject.
+	 *
+	 * @param string $certificate X509 certificate encoded as PEM
+	 * @return ?string Email address of the subject or null if it has none
+	 *
+	 * @throws ServiceException If the certificate can't be parsed
+	 */
+	public function extractEmailAddress(string $certificate): ?string {
+		$certificateData = openssl_x509_parse($certificate);
+		if ($certificateData === false) {
+			throw new ServiceException('Could not parse certificate');
+		}
+
+		return $certificateData['subject']['emailAddress'] ?? null;
+	}
+
+	/**
+	 * Check if the given private key corresponds to the given certificate.
+	 *
+	 * @param string $certificate X509 certificate encoded as PEM
+	 * @param string $privateKey Private key encoded as PEM
+	 * @return bool True if the private key matches the certificate, false otherwise or if the private key is protected by a passphrase
+	 */
+	public function checkPrivateKey(string $certificate, string $privateKey): bool {
+		return openssl_x509_check_private_key($certificate, $privateKey);
+	}
+
+	/**
+	 * Decrypt an S/MIME certificate entity that was fetched from the DB.
+	 *
+	 * @param SMimeCertificate $certificate Certificate entity from the DB
+	 * @return void
+	 *
+	 * @throws ServiceException If the decryption failed or the HMAC does not match
+	 */
+	public function decryptCertificateEntity(SMimeCertificate $certificate): void {
+		try {
+			$certificate->setCertificate($this->crypto->decrypt($certificate->getCertificate()));
+			if ($certificate->getPrivateKey() !== null) {
+				$certificate->setPrivateKey($this->crypto->decrypt($certificate->getPrivateKey()));
+			}
+		} catch (\Exception $e) {
+			throw new ServiceException('Could not decrypt certificate or private key');
```

Comment:
```suggestion
			throw new ServiceException('Could not decrypt certificate or private key: ' . $e->getMessage(), 0, $e);
```

makes debugging easier

Reply:
I have much to learn :D